### ISSUE TITLE
Maui Binding does not respect Binding.DoNothing returned from IValueConverter

### ISSUE THREAD

**Issue by Larhei** [dotnet/maui#11740]
### Description

When using a IValueConverter in a Binding that return Binding.DoNothing in Convert or ConvertBack the Data gets transferred to the source or target of the binding.

In MultiBinding this is respected. The multi binding does not update source or target in case of Binding.DoNothing.

### Steps to Reproduce

1. Checkout the provides Sample
2. Run the Sample
3. The Entry in the View should be Empty because the converter returned Binding.DoNothing. But it contains System.Object

### Link to public reproduction project repository

https://github.com/Larhei/Maui-Issues/tree/main/BindingDoNothing

### Version with bug

7.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android, Windows, macOS, Other (Tizen, Linux, etc. not supported by Microsoft directly)

### Affected platform versions

All

### Did you find any workaround?

No

**Comment by PureWeen** [comment]
@StephaneDelcroix thoughts?

**Comment by bakerhillpins** [comment]
I don't think that #11833 is related to this issue as it is compiletime, where here we have a runtime issue.

WRT this issue:
The Binding class doesn't actually have any tests for `Binding.DoNothing` or `BindableProperty.UnsetValue` (using the `FallbackValue`) being returned from the IValueConverter.  I'm actually testing code changes for this issue now. Hope to supply a PR shortly.

**Comment by mattleibow** [comment]
As far as I can tell in the code, `Binding.DoNothing` is just for cases when using `IMultiValueConverter` or `MultiBinding`. 

I am not sure using `Binding.DoNothing` is what you are looking for. It is used only in multi-bindings. In all other contexts it is treated as a normal object.

Mayb you are wanting to return null or an empty string?

**Comment by mattleibow** [comment]
That documentation is for WPF and neither Xamarin.Forms or MAUI ever did anything around that.

**Comment by bakerhillpins** [comment]
> That documentation is for WPF and neither Xamarin.Forms or MAUI ever did anything around that.

MAUI has (as far as I can tell) ZERO Class/Namespace documentation out there. MS hasn't even bothered to copy over the underwhelming Xamarin docs so developers are left with what we have.

Besides that, just because Xamarin/MAUI didn't/doesn't is far from a good reason to ignore the discussion that it should. There are plenty of situations where Xamarin and now MAUI fall short of the WPF XAML experience and we shouldn't use that as an excuse to pass over functionality. 

One of the most miserable things IMHO that resulted from the fact that Xamarin started from a private company was that they felt it necessary to alter XAML/Class naming to avoid litigation as it's end result is a splintered implementation of XAML with in-congruent basic element functionality.  I had hopes that part of the MAUI path was to merge Xamarin.XAML back to something that would match WPF.XAML. Or at least come up with a Common XAML experience.

**Comment by Larhei** [comment]
It would really help if Binding is at least the same for the basic features on all platforms.
So at least in Binding, MultiBinding, FallbackValue... just in case you want to share your ViewModels around desktop, mobile... Silverlight and all that legacy code..
I´m not asking for fancy stuff here.

And yes i know it´s Desktop... God old times when we had documentation are over.... got it

**Comment by bakerhillpins** [comment]
To follow up here:

I've modified the MAUI source to support the both the `Binding.DoNothing` and the `BindableProperty.UnsetValue` functionality described in the [IValueConverter](https://learn.microsoft.com/en-us/dotnet/api/system.windows.data.ivalueconverter.convert?view=windowsdesktop-7.0#remarks) and [IMultiValueConverter](https://learn.microsoft.com/en-us/dotnet/api/system.windows.data.imultivalueconverter.convert?view=windowsdesktop-7.0&viewFallbackFrom=net-6.0#returns) docs. These changes support both the [Binding Fallbacks](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/binding-fallbacks?view=net-maui-7.0) and the [MultiBinding](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/multibinding?view=net-maui-6.0) discussions in the docs and this would then operate as WPF does.

The one omission here is it's unclear how the [StringFormat](https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/data-binding/string-formatting?view=net-maui-7.0) behavior should be handled as it's not specifically discussed anywhere in MAUI or WPF documentation that I can find. Discussions on StackOverflow indicate that WPF bindings ignore `StringFormat` during application of `TargetNullValue` or `FallbackValue` and so I'm assuming that MAUI should keep this and I've made changes to support that.

I'll put a PR in hoping that it will generate some conversation and adoption.

**Comment by bakerhillpins** [comment]
Here are my notes on current MAUI operation. Easier to look at and understand in functional groups:

### `TargetNullValue` appears to work as it would in WPF: (`value` is null and `TargetNullValue` is not null)
- `Binding`, `TemplateBinding`, and `TypedBinding` ignore `StringFormat`.
- `MultiBinding` ignores `StringFormat` if a `IMultiValueConverter` is provided and the converter returns null.
- A `MultiBinding` without a converter always has a non null `value` because it's an `object[]` thus `TargetNullValue` is not used.

### `FallbackValue` via ValueConverter returning `BindableProperty.UnsetValue`
- MAUI doesn't implement this for `Binding`, `TemplateBinding`, and `TypedBinding`.
- `MultiBinding` ignores `StringFormat` if a `IMultiValueConverter` is provided and the converter returns  `BindableProperty.UnsetValue`. However, it *stops* with `FallbackValue` as the MAUI Docs specify. WPF docs specify returning `FallbackValue ?? BindableProperty.DefaultValue` in this case.

### `FallbackValue` due to unresolved binding.
- `Binding`, `TemplateBinding`, and `TypedBinding` avoid `StringFormat` and return `FallbackValue ?? BindableProperty.DefaultValue`.
- `MultiBinding` fails to get into this scenario as it's not evaluating the unresolved binding "status" of it's binding collection.

### No action via ValueConverter returning `Binding.DoNothing`
- MAUI doesn't implement this for `Binding`, `TemplateBinding`, and `TypedBinding`.
- `MultiBinding` will actually try to apply a provided `StringFormat` in `BindingBase.GetSourceValue(object value, Type targetPropertyType)`. If the format is successful it will return string rather than performing no action. (BUG)

As you can see there are inconsistencies in MAUI such as `FallbackValue` on a `MultiBinding` never looks to the BindableProperty.DefaultValue but it does on `Binding`, `TemplateBinding`, and `TypedBinding` for unresolved bindings.

Another apparent bug with `MultiBinding` is it's setting the "i"th binding to null if the `IMultiValueConverter.ConvertBack` method returns null in that position.  The docs (MAUI/WPF) indicate that null means the conversion was not valid/supported in that case so why would we want to use the value? 

`null to indicate that the converter cannot perform the conversion or that it does not support conversion in this direction.`

**Comment by MauiUIui** [comment]
> Mayb you are wanting to return null or an empty string?

Actually, you should never do that in a converter unless those values are really meant to be returned. An example would be when your Converter receives an invalid object. Returning null or empty string is most likely not a good option. Most of the time, throwing an exception in that case is not a good reaction either (Bindings sometimes ask for the object before it's available, so null for a non-nullable object does come unexpectedly). `Binding.DoNothing` (or `BindableProperty.UnsetValue`, I don't really know the difference between the two, didn't look at the code, but both do work afaik) should be the way to go, no mater if it's MAUI, XF, WPF or whatever. Never let a converter decide, if your value is an unexpected one.

**Comment by kubaflo** [comment]
I independently reran the generated reproduction test from draft PR [#37459](https://github.com/dotnet/maui/pull/37459) against baseline `604d9de36e05a40c36ab05cdcd68f2c6286fa260`.

**Result: the test is correctly detecting the reported behavior.** `Microsoft.Maui.Controls.Core.UnitTests.Issue11740.DoNothingConverterLeavesTargetUnchanged` executed as the only selected test and failed at the intended assertion:

```text
Binding.DoNothing should leave the target value unchanged.
```

This is a managed unit test, so its execution does not use a device; the PR's separate Android reproduction evidence is from [pipeline run 14986058](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=14986058).
